File: TimeSeriesDenoising/codeChallenge.py
import numpy as np
import matplotlib.pyplot as plt

# Load the signal package
# (No need for this in Python)

N = 4000

# Generate the original signal
t = np.linspace(0, 10 * np.pi, N)
origSignal = np.sin(t) * np.linspace(-1, 1, N) + np.random.randn(N)
r = np.random.permutation(N)
nn = int(np.round(N * 0.05))
origSignal[r[:nn]] = (1 + np.random.rand(nn)) * 10
origSignal[r[-nn:]] = -(1 + np.random.rand(nn)) * 10

# Initialize cleanedSignal with origSignal
cleanedSignal = origSignal.copy()

# Remove positive noise spikes
p2r = np.where(origSignal > 5)[0]
k = 5
for i in p2r:
    start = max(0, i - k)
    end = min(N, i + k + 1)
    cleanedSignal[i] = np.median(cleanedSignal[start:end])

# Remove negative noise spikes
p2r = np.where(origSignal < -5)[0]
k = 5
for i in p2r:
    start = max(0, i - k)
    end = min(N, i + k + 1)
    cleanedSignal[i] = np.median(cleanedSignal[start:end])

# Mean-smooth
k = 150
for i in range(N):
    start = max(0, i - k)
    end = min(N, i + k + 1)
    cleanedSignal[i] = np.mean(cleanedSignal[start:end])

# Plot the original signal and the lecturer's solution
plt.figure()
plt.subplot(3, 1, 1)
plt.plot(origSignal, linewidth=2)
plt.title("Original Signal")

plt.subplot(3, 1, 2)
plt.plot(cleanedSignal, linewidth=2)
plt.title("Lecturer's Solution")

# Your Solution
outlierThreshold = 5
k = 5
myCleanedSignal = origSignal.copy()

# Positive outliers
outlierIndices = np.where(origSignal > outlierThreshold)[0]
for i in outlierIndices:
    start = max(0, i - k)
    end = min(N, i + k + 1)
    myCleanedSignal[i] = np.median(origSignal[start:end])

# Negative outliers
outlierIndices = np.where(origSignal < -outlierThreshold)[0]
for i in outlierIndices:
    start = max(0, i - k)
    end = min(N, i + k + 1)
    myCleanedSignal[i] = np.median(origSignal[start:end])

# Moving average smoothing
windowSize = 151
movingAvgFilter = np.ones(windowSize) / windowSize
mySmoothSignal = np.convolve(myCleanedSignal, movingAvgFilter, mode='same')

plt.subplot(3, 1, 3)
plt.plot(mySmoothSignal, color=[0.8, 0.2, 0.1], linewidth=2)
plt.title('Your Solution')

plt.tight_layout()
plt.show()

File: TimeSeriesDenoising/codeChallenge.m
%%
% COURSE: Signal processing and image processing in MATLAB and Python
% SECTION: Time-domain denoising
% VIDEO: Code challenge: Denoise these signals!
% Instructor: mikexcohen.com
%
%%

%% added
clc;
clear;

% Load the signal package
pkg load signal;

N = 4000;

origSignal = linspace(-1,1,N) .* sin(linspace(0,10*pi,N)) + randn(1,N);
r = randperm(N);
nn = round(N*.05);
origSignal(r(1:nn)) = (1+rand(1,nn))*10;
origSignal(r(end-nn+1:end)) = -(1+rand(1,nn))*10;

% Initialize cleanedSignal with origSignal
cleanedSignal = origSignal;

% remove positive noise spikes
p2r = find(origSignal > 5);
k = 5;
for i = 1:length(p2r)
    cleanedSignal(p2r(i)) = median(cleanedSignal(max(1, p2r(i) - k):min(N, p2r(i) + k)));
end

% remove negative noise spikes
p2r = find(origSignal < -5);
k = 5;
for i = 1:length(p2r)
    cleanedSignal(p2r(i)) = median(cleanedSignal(max(1, p2r(i) - k):min(N, p2r(i) + k)));
end

% mean-smooth
k = 150;
for i = 1:N
    cleanedSignal(i) = mean(cleanedSignal(max(1, i - k):min(N, i + k)));
end

% Ensure the figure is cleared before plotting
figure;
subplot(3,1,1);
plot(1:N, origSignal, 'linew', 2);
title("Original Signal");

subplot(3,1,2);
plot(1:N, cleanedSignal, 'linew', 2);
title("Lecturer's Solution");

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%% Your Result: %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% Load your saved data
load denoising_codeChallenge.mat;

% Ensure the size matches
N = length(origSignal);

% Step 1: Identify and replace outliers
k = 5; % Window size for median filter
outlierThreshold = 5; % Threshold for identifying outliers
myCleanedSignal = origSignal;

% Positive outliers
outlierIndices = find(origSignal > outlierThreshold);
for i = 1:length(outlierIndices)
    windowStart = max(1, outlierIndices(i) - k);
    windowEnd = min(N, outlierIndices(i) + k);
    myCleanedSignal(outlierIndices(i)) = median(origSignal(windowStart:windowEnd));
end

% Negative outliers
outlierIndices = find(origSignal < -outlierThreshold);
for i = 1:length(outlierIndices)
    windowStart = max(1, outlierIndices(i) - k);
    windowEnd = min(N, outlierIndices(i) + k);
    myCleanedSignal(outlierIndices(i)) = median(origSignal(windowStart:windowEnd));
end

% Step 2: Moving average smoothing
windowSize = 150; % Window size for moving average (odd number)
movingAvgFilter = ones(1, windowSize) / windowSize; % Normalized moving average filter
mySmoothSignal = conv(myCleanedSignal, movingAvgFilter, 'same'); % Apply moving average filter

subplot(3,1,3);
plot(1:N, mySmoothSignal, 'Color', [0.8 0.2 0.1], 'linew', 2);
title('Your Solution');

%% done.

